homeassistant/auth/providers/openid.py

```diff
@@ -0,0 +1,294 @@
+"""OpenId based authentication provider."""
+
+import logging
+import re
+import secrets
+from typing import Any, Dict, Optional, cast
+
+from aiohttp import ClientResponseError
+from aiohttp.client import ClientResponse
+from aiohttp.web import HTTPBadRequest, Request, Response
+from jose import jwt
+import voluptuous as vol
+from yarl import URL
+
+from homeassistant.components.http import HomeAssistantView
+from homeassistant.core import HomeAssistant
+from homeassistant.exceptions import HomeAssistantError
+from homeassistant.helpers.aiohttp_client import async_get_clientsession
+from homeassistant.helpers.config_entry_oauth2_flow import _decode_jwt, _encode_jwt
```

Comment:
This is somewhat ugly. I'm re-using the jwt encoding for flows form the oauth2 helper. Sadly i can't re-use the whole function/view currently. It is locked to config flow.